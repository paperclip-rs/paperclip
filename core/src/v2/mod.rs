//! Core types and traits associated with the
//! [OpenAPI v2 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).

#[cfg(feature = "actix-base")]
mod actix;
mod extensions;
pub mod models;
#[cfg(feature = "codegen")]
mod resolver;
pub mod schema;

#[cfg(feature = "actix-base")]
pub use self::actix::{
    AcceptedJson, CreatedJson, NoContent, OperationModifier, ResponderWrapper, ResponseWrapper,
};

pub use self::models::{DefaultSchema, ResolvableApi};
pub use self::schema::Schema;
pub use paperclip_macros::*;

#[cfg(feature = "codegen")]
use self::resolver::Resolver;
#[cfg(feature = "codegen")]
use crate::error::ValidationError;

#[cfg(feature = "codegen")]
impl<S: Schema + Default> ResolvableApi<S> {
    /// Consumes this API schema, resolves the references and returns
    /// the resolved schema.
    ///
    /// This walks recursively, collects the referenced schema objects,
    /// substitutes the referenced IDs with the pointer to schema objects
    /// and returns the resolved object or an error if it encountered one.
    pub fn resolve(self) -> Result<ResolvableApi<S>, ValidationError> {
        let mut resolver = Resolver::from((
            self.definitions,
            self.paths,
            self.parameters,
            self.responses,
        ));
        resolver.resolve()?;
        Ok(ResolvableApi {
            swagger: self.swagger,
            info: self.info,
            definitions: resolver.defs,
            paths: resolver.paths,
            base_path: self.base_path,
            host: self.host,
            schemes: self.schemes,
            consumes: self.consumes,
            produces: self.produces,
            coders: self.coders,
            support_crates: self.support_crates,
            parameters: resolver.params,
            responses: resolver.resp,
            spec_format: self.spec_format,
            external_docs: self.external_docs,
            security: self.security,
            security_definitions: self.security_definitions,
            tags: self.tags,
        })
    }
}
