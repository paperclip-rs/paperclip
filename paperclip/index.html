<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Paperclip is a OpenAPI code generator for efficient type-safe compile-time checked HTTP APIs in Rust."><title>paperclip - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="paperclip" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0-nightly (f3db63916 2025-06-17)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../paperclip/index.html">paperclip</a><span class="version">0.9.6</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#modules">Crate Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>paperclip</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/paperclip/lib.rs.html#1-34">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Paperclip is a OpenAPI code generator for efficient type-safe
compile-time checked HTTP APIs in Rust.</p>
<p>See the <a href="https://paperclip-rs.github.io/paperclip">website</a> for detailed
documentation and examples.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="actix/index.html" title="mod paperclip::actix">actix</a></dt><dd>Plugin types, traits and macros for actix-web framework.</dd><dt><a class="mod" href="util/index.html" title="mod paperclip::util">util</a></dt><dt><a class="mod" href="v2/index.html" title="mod paperclip::v2">v2</a></dt><dd>Utilities related to the <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md">OpenAPI v2 specification</a>.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.PaperClipError.html" title="enum paperclip::PaperClipError">Paper<wbr>Clip<wbr>Error</a></dt><dd>Global error which encapsulates all related errors.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="type" href="type.PaperClipResult.html" title="type paperclip::PaperClipResult">Paper<wbr>Clip<wbr>Result</a></dt><dd>Generic result used throughout this library.</dd></dl><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="attr" href="attr.api_v2_schema.html" title="attr paperclip::api_v2_schema">api_<wbr>v2_<wbr>schema</a></dt><dd>Converts your struct to support deserializing from an OpenAPI v2
<a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject">Schema</a>
object (<a href="https://paperclip-rs.github.io/paperclip/paperclip/v2/">example</a>). This adds the necessary fields (in addition to your own fields) and implements the
<code>Schema</code> trait for parsing and codegen.</dd></dl></section></div></main></body></html>