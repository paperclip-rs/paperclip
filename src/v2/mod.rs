//! Utilities related to the [OpenAPI v2 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
//!
//! # Detailed example
//!
//! To parse your v2 spec, you begin with transforming the schema into a
//! Rust struct. If your schema doesn't have custom properties, then you
//! can use the `DefaultSchema`.
//!
//! ```rust,no_run
//! use paperclip::v2::{self, ResolvableApi, DefaultSchema, models::Version};
//!
//! use std::fs::File;
//!
//! let mut fd = File::open("my_spec.yaml").unwrap(); // yaml or json
//! let api: ResolvableApi<DefaultSchema> = v2::from_reader(&mut fd).unwrap();
//! assert_eq!(api.swagger, Version::V2);
//! ```
//!
//! On the other hand, if your schema does have custom properties which you'd
//! like to parse, then use the `#[api_v2_schema]` proc macro.
//!
//! For example, let's take the [Kubernetes API spec][kube-spec]
//! which uses some custom thingmabobs. Let's say we're only interested in the
//! `x-kubernetes-patch-strategy` field for now.
//!
//! [kube-spec]: https://github.com/kubernetes/kubernetes/tree/afd928b8bc81cea385eba4c94558373df7aeae75/api/openapi-spec
//!
//! ```rust,no_run
//! #[macro_use] extern crate paperclip;
//! #[macro_use] extern crate serde_derive; // NOTE: We're using serde for decoding stuff.
//!
//! use paperclip::v2::{self, ResolvableApi};
//!
//! use std::fs::File;
//!
//! #[derive(Debug, Deserialize)]
//! #[serde(rename_all = "camelCase")]
//! enum PatchStrategy {
//!     Merge,
//!     RetainKeys,
//!     #[serde(rename = "merge,retainKeys")]
//!     MergeAndRetain,
//!     #[serde(other)]
//!     Other,
//! }
//!
//! #[api_v2_schema]
//! #[derive(Debug, Deserialize)]
//! struct K8sSchema {
//!     #[serde(rename = "x-kubernetes-patch-strategy")]
//!     patch_strategy: Option<PatchStrategy>,
//! }
//!
//! // K8sSchema now implements `Schema` trait.
//! let mut fd = File::open("k8s_spec.yaml").unwrap();
//! let api: ResolvableApi<K8sSchema> = v2::from_reader(&mut fd).unwrap();
//! ```
//!
//! Now, if `codegen` feature is enabled (it is by default), we can use the
//! emitter to emit the API into some path. But first, we need to resolve the
//! raw schema. During resolution, we:
//!
//! - walk through the nodes, find `$ref` fields and assign references to
//! the actual definitions.
//! - identify anonymous definitions in body parameters and response schemas
//! and add them to the known map of definitions.
//!
//! ```rust,no_run
//! #[cfg(feature = "codegen")] {
//! # use paperclip::v2::{self, ResolvableApi, DefaultSchema};
//! # let api: ResolvableApi<DefaultSchema> = v2::from_reader(&mut std::io::Cursor::new(vec![])).unwrap();
//!
//! let resolved = api.resolve().unwrap();
//! }
//! ```
//!
//! ```rust,no_run
//! #[cfg(feature = "codegen")] {
//! # use paperclip::v2::{self, ResolvableApi, DefaultSchema};
//! # let api: ResolvableApi<DefaultSchema> = v2::from_reader(&mut std::io::Cursor::new(vec![])).unwrap();
//! use paperclip::v2::{DefaultEmitter, EmitterState, Emitter};
//!
//! let mut state = EmitterState::default();
//! state.working_dir = "/path/to/my/crate".into();
//! let emitter = DefaultEmitter::from(state);
//! emitter.generate(&api).unwrap(); // generate code!
//! }
//! ```

#[cfg(feature = "codegen")]
pub mod codegen;

use crate::error::PaperClipError;
use paperclip_core::v2::models::SpecFormat;
use serde::Deserialize;

use std::io::{Read, Seek, SeekFrom};

#[cfg(feature = "codegen")]
pub use self::codegen::{DefaultEmitter, Emitter, EmitterState};
pub use paperclip_core::{
    im,
    v2::{
        models::{self, DefaultSchema, ResolvableApi},
        schema::{self, Schema},
    },
};

/// Deserialize the schema from the given reader. Currently, this only supports
/// JSON and YAML formats.
pub fn from_reader<R, S>(mut reader: R) -> Result<ResolvableApi<S>, PaperClipError>
where
    R: Read + Seek,
    for<'de> S: Deserialize<'de> + Schema,
{
    let mut buf = [0; 1];
    reader.read_exact(&mut buf)?;
    reader.seek(SeekFrom::Start(0))?;

    // FIXME: Support whitespaces
    let (mut api, fmt) = if buf[0] == b'{' {
        (
            serde_json::from_reader::<_, ResolvableApi<S>>(reader)?,
            SpecFormat::Json,
        )
    } else {
        (
            serde_yaml::from_reader::<_, ResolvableApi<S>>(reader)?,
            SpecFormat::Yaml,
        )
    };

    api.spec_format = fmt;
    Ok(api)
}
