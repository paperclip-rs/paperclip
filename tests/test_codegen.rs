use paperclip::v2::{
    self,
    codegen::{CrateMeta, DefaultEmitter, EmitMode, Emitter, EmitterState},
    models::{DefaultSchema, ResolvableApi},
};

use once_cell::sync::Lazy;
use std::{fs::File, path::PathBuf};

static ROOT: Lazy<PathBuf> = Lazy::new(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")));
static PET_SCHEMA: Lazy<ResolvableApi<DefaultSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/pet-v2.yaml")).expect("file?");
    let raw: ResolvableApi<DefaultSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static K8S_SCHEMA: Lazy<ResolvableApi<K8sSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/k8s-v1.16.0-alpha.0-openapi-v2.json")).expect("file?");
    let raw: ResolvableApi<K8sSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static CODEGEN_PET_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_LIB_NO_ROOT: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet/no_root");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    meta.no_root = true;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_pet/cli");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});

static CODEGEN_K8S_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests");
    state.working_dir.push("test_k8s");
    state.mod_prefix = "crate::codegen::";

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});
static CODEGEN_K8S_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_k8s/cli");
    let mut meta = CrateMeta::default();
    assert_eq!(meta.mode, EmitMode::Module);
    meta.name = Some("test-k8s-cli".into());
    meta.version = Some("0.0.0".into());
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
enum PatchStrategy {
    Merge,
    RetainKeys,
    #[serde(rename = "merge,retainKeys")]
    MergeAndRetain,
}

#[paperclip::api_v2_schema]
#[allow(dead_code)]
#[derive(Debug, serde::Deserialize)]
struct K8sSchema {
    #[serde(rename = "x-kubernetes-patch-strategy")]
    patch_strategy: Option<PatchStrategy>,
}

static CODEGEN: Lazy<()> = Lazy::new(|| {
    env_logger::builder()
        .filter(Some("paperclip"), log::LevelFilter::Info)
        .init();
    Lazy::force(&CODEGEN_PET_LIB);
    Lazy::force(&CODEGEN_PET_LIB_NO_ROOT);
    Lazy::force(&CODEGEN_PET_CLI);
    Lazy::force(&CODEGEN_K8S_LIB);
    Lazy::force(&CODEGEN_K8S_CLI);
});

fn assert_file(path: &str) {
    Lazy::force(&CODEGEN);

    let data = std::fs::read_to_string(ROOT.join(path))
        .unwrap_or_else(|err| panic!("missing file {}: {}", path, err));
    insta::assert_snapshot!(path, data);
}

#[cfg(test)]
mod tests_pet {
    use super::*;
    #[test]
    fn test_lib_root() {
        assert_file("tests/test_pet/lib.rs");
    }

    #[test]
    fn test_no_root_manifest() {
        assert_file("tests/test_pet/no_root/Cargo.toml");
    }

    #[test]
    fn test_manifest() {
        assert_file("tests/test_pet/Cargo.toml");
    }

    #[test]
    fn test_header_parameters() {
        assert_file("tests/test_pet/pet.rs");
    }

    #[test]
    fn test_cli_params() {
        assert_file("tests/test_pet/cli/app.yaml");
    }

    #[test]
    fn test_recusrive_object() {
        assert_file("tests/test_pet/recursive_container.rs");
        assert_file("tests/test_pet/recursive_object.rs");
    }

    #[test]
    fn test_anonymous_object_definition_in_schema() {
        // An object "can" define objects in its schema without referencing
        // them from known definitions. In that case, we autogenerate stuff.
        assert_file("tests/test_pet/order.rs");
    }
    #[test]
    fn test_anonymous_object_definition_in_body() {
        assert_file("tests/test_pet/post_shipments_body.rs");
    }

    #[test]
    fn test_anonumous_object_definition_in_response() {
        assert_file("tests/test_pet/get_shipments_id_response.rs");
    }

    #[test]
    fn test_simple_array_parameter_in_path() {
        assert_file("tests/test_pet/status.rs");
    }

    #[test]
    fn test_misc() {
        assert_file("tests/test_pet/miscellaneous.rs");
    }

    #[test]
    fn test_multipart_with_file() {
        assert_file("tests/test_pet/cli/status.rs");
    }

    #[test]
    fn test_simple_any_enum() {
        assert_file("tests/test_pet/test_enum.rs");
    }
}

#[cfg(test)]
mod tests_k8s {
    use super::*;
    use paperclip::v2::models::{HttpMethod, Version};

    #[test]
    fn test_definition_ref_cycles() {
        assert_eq!(K8S_SCHEMA.swagger, Version::V2);
        assert_eq!(K8S_SCHEMA.definitions.len(), 663);

        let json_props_def = &K8S_SCHEMA.definitions
            ["io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps"];
        let desc = json_props_def.read().description.clone();
        let all_of = json_props_def.read().properties["allOf"].clone();
        let items = all_of.read().items.as_ref().unwrap().clone();
        assert_eq!(items.read().description, desc); // both point to same `JSONSchemaProps`
    }

    #[test]
    fn test_resolved_schema() {
        let resp = &K8S_SCHEMA.paths["/api/"].methods[&HttpMethod::Get].responses["200"].read();
        let schema = resp.schema.as_ref().expect("bleh?").read();
        assert!(schema.reference.is_none()); // this was a reference
        assert_eq!(
            &K8S_SCHEMA.definitions["io.k8s.apimachinery.pkg.apis.meta.v1.APIVersions"]
                .read()
                .description
                .as_ref()
                .unwrap()
                .as_str(),
            schema.description.as_ref().unwrap()
        );
    }

    #[test]
    fn test_child_module_declarations() {
        assert_file("tests/test_k8s/io/k8s/api/mod.rs");
    }

    #[test]
    fn test_transparency_with_params() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/custom_resource_definition.rs");
    }
    #[test]
    fn test_struct_for_complex_object() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/json_schema_props.rs");
    }

    #[test]
    fn test_root_mod() {
        assert_file("tests/test_k8s/mod.rs");
    }

    #[test]
    fn test_generics_mod() {
        assert_file("tests/test_k8s/generics.rs");
    }

    #[test]
    fn test_same_object_creates_multiple_builders() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/config_map.rs")
    }

    #[test]
    fn test_same_object_with_multiple_builders_has_basic_builder() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/pod.rs")
    }

    #[test]
    fn test_simple_object_builder_with_required_fields() {
        assert_file("tests/test_k8s/io/k8s/api/rbac/v1/policy_rule.rs")
    }

    #[test]
    fn test_builder_with_field_parameter_collision_and_method_collision() {
        // grace_period_seconds, orphan_dependents and propagation_policy
        // exist in both the object and as a query parameter. If one is set,
        // we should also set the other.
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/delete_options.rs");
    }

    #[test]
    fn test_unit_builder_with_no_modifier() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/api_group_list.rs");
    }

    #[test]
    fn test_builder_field_with_iterators() {
        assert_file(
            "tests/test_k8s/io/k8s/api/certificates/v1beta1/certificate_signing_request_spec.rs",
        );
    }

    #[test]
    fn test_any_in_operation_bound_to_unrelated_struct() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/patch.rs");
    }

    #[test]
    fn test_cli_manifest() {
        assert_file("tests/test_k8s/cli/Cargo.toml");
    }

    #[test]
    fn test_cli_main() {
        assert_file("tests/test_k8s/cli/main.rs");
    }

    #[test]
    fn test_clap_yaml() {
        assert_file("tests/test_k8s/cli/app.yaml");
    }
}
