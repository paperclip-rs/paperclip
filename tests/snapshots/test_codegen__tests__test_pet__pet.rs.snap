---
source: tests/test_codegen.rs
expression: data
---

/// A pet is a person's best friend
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Pet<Any> {
    pub category: Option<crate::category::Category>,
    pub id: i64,
    pub name: String,
    #[serde(rename = "photoUrls")]
    pub photo_urls: Option<Vec<String>>,
    pub tags: Option<Vec<crate::tag::Tag>>,
    #[serde(flatten)]
    pub other_fields: Option<paperclip::v2::PropertiesMap<String, Any>>,
}

impl<Any: Default> Pet<Any> {
    /// Create a builder for this object.
    #[inline]
    pub fn builder() -> PetBuilder<crate::generics::MissingId, crate::generics::MissingName, Any> {
        PetBuilder {
            body: Default::default(),
            _id: core::marker::PhantomData,
            _name: core::marker::PhantomData,
        }
    }

    /// Fetch list of pets
    #[inline]
    pub fn list_pets() -> PetGetBuilder {
        PetGetBuilder
    }

    /// Add a new pet to the store
    #[inline]
    pub fn add_pet() -> PetPostBuilder<crate::generics::MissingXAuth, crate::generics::MissingId, crate::generics::MissingName, Any> {
        PetPostBuilder {
            inner: Default::default(),
            _param_x_auth: core::marker::PhantomData,
            _id: core::marker::PhantomData,
            _name: core::marker::PhantomData,
        }
    }

    /// Find pet by ID
    #[inline]
    pub fn get_pet_by_id() -> PetGetBuilder1<crate::generics::MissingPetId> {
        PetGetBuilder1 {
            inner: Default::default(),
            _param_pet_id: core::marker::PhantomData,
        }
    }
}

impl<Any> Into<Pet<Any>> for PetBuilder<crate::generics::IdExists, crate::generics::NameExists, Any> {
    fn into(self) -> Pet<Any> {
        self.body
    }
}

impl<Any> Into<Pet<Any>> for PetPostBuilder<crate::generics::XAuthExists, crate::generics::IdExists, crate::generics::NameExists, Any> {
    fn into(self) -> Pet<Any> {
        self.inner.body
    }
}

/// Builder for [`Pet`](./struct.Pet.html) object.
#[derive(Debug, Clone)]
pub struct PetBuilder<Id, Name, Any> {
    body: self::Pet<Any>,
    _id: core::marker::PhantomData<Id>,
    _name: core::marker::PhantomData<Name>,
}

impl<Id, Name, Any> PetBuilder<Id, Name, Any> {
    #[inline]
    pub fn category(mut self, value: crate::category::Category) -> Self {
        self.body.category = Some(value.into());
        self
    }

    #[inline]
    pub fn id(mut self, value: impl Into<i64>) -> PetBuilder<crate::generics::IdExists, Name, Any> {
        self.body.id = value.into();
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn name(mut self, value: impl Into<String>) -> PetBuilder<Id, crate::generics::NameExists, Any> {
        self.body.name = value.into();
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn photo_urls(mut self, value: impl Iterator<Item = impl Into<String>>) -> Self {
        self.body.photo_urls = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn tags(mut self, value: impl Iterator<Item = crate::tag::Tag>) -> Self {
        self.body.tags = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn other_fields(mut self, value: impl Iterator<Item = (String, impl Into<Any>)>) -> Self {
        self.body.other_fields = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }
}

/// Builder created by [`Pet::list_pets`](./struct.Pet.html#method.list_pets) method for a `GET` operation associated with `Pet`.
#[derive(Debug, Clone)]
pub struct PetGetBuilder;


impl<Client: crate::client::ApiClient + Sync + 'static> crate::client::Sendable<Client> for PetGetBuilder {
    type Output = Vec<Pet<serde_yaml::Value>>;

    const METHOD: http::Method = http::Method::GET;

    fn rel_path(&self) -> std::borrow::Cow<'static, str> {
        "/pets".into()
    }
}

/// Builder created by [`Pet::add_pet`](./struct.Pet.html#method.add_pet) method for a `POST` operation associated with `Pet`.
#[repr(transparent)]
#[derive(Debug, Clone)]
pub struct PetPostBuilder<XAuth, Id, Name, Any> {
    inner: PetPostBuilderContainer<Any>,
    _param_x_auth: core::marker::PhantomData<XAuth>,
    _id: core::marker::PhantomData<Id>,
    _name: core::marker::PhantomData<Name>,
}

#[derive(Debug, Default, Clone)]
struct PetPostBuilderContainer<Any> {
    body: self::Pet<Any>,
    param_x_auth: Option<String>,
    param_x_pet_id: Option<i64>,
}

impl<XAuth, Id, Name, Any> PetPostBuilder<XAuth, Id, Name, Any> {
    #[inline]
    pub fn x_auth(mut self, value: impl Into<String>) -> PetPostBuilder<crate::generics::XAuthExists, Id, Name, Any> {
        self.inner.param_x_auth = Some(value.into());
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn x_pet_id(mut self, value: impl Into<i64>) -> Self {
        self.inner.param_x_pet_id = Some(value.into());
        self
    }

    #[inline]
    pub fn category(mut self, value: crate::category::Category) -> Self {
        self.inner.body.category = Some(value.into());
        self
    }

    #[inline]
    pub fn id(mut self, value: impl Into<i64>) -> PetPostBuilder<XAuth, crate::generics::IdExists, Name, Any> {
        self.inner.body.id = value.into();
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn name(mut self, value: impl Into<String>) -> PetPostBuilder<XAuth, Id, crate::generics::NameExists, Any> {
        self.inner.body.name = value.into();
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn photo_urls(mut self, value: impl Iterator<Item = impl Into<String>>) -> Self {
        self.inner.body.photo_urls = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn tags(mut self, value: impl Iterator<Item = crate::tag::Tag>) -> Self {
        self.inner.body.tags = Some(value.map(|value| value.into()).collect::<Vec<_>>().into());
        self
    }

    #[inline]
    pub fn other_fields(mut self, value: impl Iterator<Item = (String, impl Into<Any>)>) -> Self {
        self.inner.body.other_fields = Some(value.map(|(key, value)| (key, value.into())).collect::<paperclip::v2::PropertiesMap<_, _>>().into());
        self
    }
}

impl<Client: crate::client::ApiClient + Sync + 'static, Any: serde::Serialize> crate::client::Sendable<Client> for PetPostBuilder<crate::generics::XAuthExists, crate::generics::IdExists, crate::generics::NameExists, Any> {
    type Output = crate::pet::Pet<serde_yaml::Value>;

    const METHOD: http::Method = http::Method::POST;

    fn rel_path(&self) -> std::borrow::Cow<'static, str> {
        "/pets".into()
    }

    fn modify(&self, req: Client::Request) -> Result<Client::Request, crate::client::ApiError<Client::Response>> {
        use crate::client::Request;
        let mut req = req;
        req = req.header("X-Auth", &self.inner.param_x_auth.as_ref().map(std::string::ToString::to_string).expect("missing parameter x_auth?"));
        if let Some(v) = &self.inner.param_x_pet_id.as_ref().map(std::string::ToString::to_string) {
            req = req.header("X-Pet-ID", &v);
        }

        Ok(req
        .header(http::header::CONTENT_TYPE.as_str(), "application/yaml")
        .body_bytes({
            let mut vec = vec![];
            serde_yaml::to_writer(&mut vec, &self.inner.body)?;
            vec
        })
        .header(http::header::ACCEPT.as_str(), "application/yaml"))
    }
}

impl<Any> crate::client::ResponseWrapper<crate::pet::Pet<serde_yaml::Value>, PetPostBuilder<crate::generics::XAuthExists, crate::generics::IdExists, crate::generics::NameExists, Any>> {
    /// Maximum allowed requests in the current period
    #[inline]
    pub fn x_rate_limit(&self) -> Option<i64> {
        self.headers.get("x-rate-limit").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
    /// Whether the requests have exceeded for this window.
    #[inline]
    pub fn x_rate_limit_exceeded(&self) -> Option<bool> {
        self.headers.get("x-rate-limit-exceeded").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
    /// Remaining requests in the current period
    #[inline]
    pub fn x_rate_limit_remaining(&self) -> Option<i64> {
        self.headers.get("x-rate-limit-remaining").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
    /// Time at which rate limit is reset (in UNIX epoch)
    #[inline]
    pub fn x_rate_limit_reset(&self) -> Option<i64> {
        self.headers.get("x-rate-limit-reset").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
    #[inline]
    pub fn x_array(&self) -> Option<crate::util::Delimited<crate::util::Delimited<crate::util::Delimited<crate::util::Delimited<f64, crate::util::Ssv>, crate::util::Tsv>, crate::util::Csv>, crate::util::Csv>> {
        self.headers.get("x-array").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
    #[inline]
    pub fn x_string(&self) -> Option<String> {
        self.headers.get("x-string").and_then(|v| String::from_utf8_lossy(v.as_ref()).parse().ok())
    }
}

/// Builder created by [`Pet::get_pet_by_id`](./struct.Pet.html#method.get_pet_by_id) method for a `GET` operation associated with `Pet`.
#[repr(transparent)]
#[derive(Debug, Clone)]
pub struct PetGetBuilder1<PetId> {
    inner: PetGetBuilder1Container,
    _param_pet_id: core::marker::PhantomData<PetId>,
}

#[derive(Debug, Default, Clone)]
struct PetGetBuilder1Container {
    param_pet_id: Option<i64>,
}

impl<PetId> PetGetBuilder1<PetId> {
    /// ID of the pet.
    #[inline]
    pub fn pet_id(mut self, value: impl Into<i64>) -> PetGetBuilder1<crate::generics::PetIdExists> {
        self.inner.param_pet_id = Some(value.into());
        unsafe { std::mem::transmute(self) }
    }
}

impl<Client: crate::client::ApiClient + Sync + 'static> crate::client::Sendable<Client> for PetGetBuilder1<crate::generics::PetIdExists> {
    type Output = Pet<serde_yaml::Value>;

    const METHOD: http::Method = http::Method::GET;

    fn rel_path(&self) -> std::borrow::Cow<'static, str> {
        format!("/pets/{petId}", petId=self.inner.param_pet_id.as_ref().expect("missing parameter pet_id?")).into()
    }
}

